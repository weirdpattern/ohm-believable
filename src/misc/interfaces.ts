/**
 * @typedef {Function} OnEditBand
 * @param {string | null} band the band whose value changed.
 * @returns {void}
 */
type OnEditBandHandler = (band: Maybe<keyof ResistorBands>) => void;

/**
 * @typedef {Function} OnEditBandTimeoutHandler
 * @returns {void}
 */
type OnEditBandTimeoutHandler = () => void;

/**
 * @typedef {Function} OnColorChangeHandler
 * @param {string} band the band whose value changed.
 * @param {string} color the new color of the band.
 * @returns {void}
 */
type OnColorChangeHandler = (
  band: keyof ResistorBands,
  color: AllColors
) => void;

/**
 * @typedef {Function} OnColorPickerChangeHandler
 * @param {string} color the new color of the band.
 * @returns {void}
 */
type OnColorPickerChangeHandler = (color: string) => void;

/**
 * @typedef {Type} Maybe<T>
 */
export type Maybe<T> = T | null;

/**
 * @typedef {Type} BaseColors
 * Applies to all types of bands
 */
export type BaseColors =
  | "brown"
  | "red"
  | "yellow"
  | "green"
  | "blue"
  | "violet"
  | "gray";

/**
 * @typedef {Type} BandColors
 * Applies to bands A and B.
 */
export type BandColors = BaseColors | "black" | "white";

/**
 * @typedef {Type} MultiplierColors
 * Applies to multiplier.
 */
export type MultiplierColors = BandColors | "pink" | "silver" | "gold";

/**
 * @typedef {Type} ToleranceColors
 * Applies to tolerance.
 */
export type ToleranceColors = BaseColors | "silver" | "gold" | "transparent";

/**
 * @typedef {Type} AllColors
 */
export type AllColors = BandColors | MultiplierColors | ToleranceColors;

/**
 * Describes the bands.
 * @typedef {Interface} ResistorBands
 * @property {BandColors} a the a band.
 * @property {BandColors} b the b band.
 * @property {MultiplierColors} multiplier the resistor multipler.
 * @property {ToleranceColors} tolerance the tolerance of the resistor.
 */
export interface ResistorBands {
  a: BandColors;
  b: BandColors;
  multiplier: MultiplierColors;
  tolerance: ToleranceColors;
}

/**
 * Describes the application state.
 * @typedef {Interface} ApplicationState
 * @property {ResistorBands} bands the bands.
 * @property {string | null} editingBand the band being edited.
 */
export interface ApplicationState {
  bands: ResistorBands;
  editingBand: Maybe<keyof ResistorBands>;
}

/**
 * Describes the bands properties.
 * @typedef {ApplicationState} BandsProps
 * @property {ResistorBands} bands the bands.
 * @property {string | null} editingBand the band being edited.
 * @property {OnEditBandHandler} onEditBand the edit band handler to be used.
 * @property {OnColorChangeHandler} onColorChange
 *    the color change handler to be used.
 */
export interface BandsProps extends ApplicationState {
  onEditBand: OnEditBandHandler;
  onColorChange: OnColorChangeHandler;
}

/**
 * Describes the color picker properties.
 * @typedef {Interface} ColorPickerProps
 * @property {string[]} colors the colors supported by the picker.
 * @property {OnColorChangeHandler} onColorChange
 *    the color change handler to be used.
 */
export interface ColorPickerProps {
  colors: string[];
  onColorChange: OnColorPickerChangeHandler;
  onEditBandTimeout: OnEditBandTimeoutHandler;
}

/**
 * Describes the resistance properties.
 * @typedef {ResistorBands} ResistanceProps
 * @property {BandColors} a the a band.
 * @property {BandColors} b the b band.
 * @property {MultiplierColors} multiplier the resistor multipler.
 * @property {ToleranceColors} tolerance the tolerance of the resistor.
 */
export interface ResistanceProps extends ResistorBands {}

/**
 * Describes a color descriptor.
 * @typedef {Interface} ColorDescriptor
 * @property {string} color the name of the color.
 * @property {string} hex the hexadecimal color representation.
 * @property {number} [value] the significant value of the color.
 * @property {number} [multiplier] the multiplier of the color.
 * @property {string} [tolerance] the tolerance of the color.
 */
export interface ColorDescriptor {
  color: string;
  hex: string;
  value: Maybe<number>;
  multiplier: Maybe<number>;
  tolerance: Maybe<string>;
}

/**
 * Describes an option descriptor.
 * @typedef {Interface} OptionDescriptor
 * @property {string} text the label to be displayed.
 * @property {number} value the value to be used.
 * @property {string} cssClass the css class to be used.
 */
export interface OptionDescriptor {
  text: string;
  value: number;
  cssClass: string;
}

/**
 * Determines if value is null on a Maybe type.
 * @param {Maybe<T>} value the value to be validated.
 * @returns {boolean} true when value is null; false otherwise.
 */
export function isMaybe<T>(value: Maybe<T>): value is null {
  return value === null;
}
