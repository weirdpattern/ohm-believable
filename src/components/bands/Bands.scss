$color-background: #E9DDAF;

$colors: (
  black: #000000,
  brown: #964B00,
  red: #FF0000,
  orange: #FFA500,
  yellow: #FFFF00,
  green: #9ACD32,
  blue: #6495ED,
  violet: #9400D3,
  gray: #A0A0A0,
  white: #FFFFFF,
  gold: #CFB53B,
  silver: #C0C0C0,
  pink: #FF69B4,
  transparent: $color-background
);

.resistor-svg {
  width: 100%;
  text-align: center;
  margin-top: 30px;

  .body {
    fill: $color-background;
    stroke-width: 0.5;
    stroke-linecap: butt;
    stroke-linejoin: miter;
    stroke-opacity: 1;

    .clickable {
      opacity: 1;
      fill-opacity: 1;
      stroke: none;
      stroke-width: 0.94488192;
      stroke-opacity: 1;
    }

    .line {
      color: map-get($colors, black);
    }
  }
}

@each $color, $value in $colors {
  .#{$color} {
    @if $color == transparent {
      fill: $color-background;
      stroke: #000000;
      stroke-dasharray: 5, 5;
    } @else {
      fill: $value;
    }
  }
}
