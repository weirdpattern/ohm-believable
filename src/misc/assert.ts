/**
 * Asserts a given condition.
 * @param {boolean} condition the condition to be validated.
 * @param {string} message the message to be displayed.
 * @returns {void | never} throws when condition is not met.
 */
export default (condition: boolean, message: string): void | never => {
  if (!condition) {
    throw new Error(message);
  }
};
