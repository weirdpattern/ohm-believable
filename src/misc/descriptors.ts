import { ColorDescriptor, Maybe } from "./interfaces";

/**
 * Creates a new ColorDescriptor.
 * @param {string} color the name of the color.
 * @param {string} hex the hexadecimal color representation.
 * @param {number} [value] the significant value of the color.
 * @param {number} multiplier the multiplier of the color.
 * @param {string} [tolerance] the tolerance of the color.
 * @returns {ColorDescriptor} the descriptor.
 */
function descriptor(
  color: string,
  hex: string,
  value: Maybe<number>,
  multiplier: Maybe<number>,
  tolerance: Maybe<string>
): ColorDescriptor {
  return {
    color,
    hex,
    value,
    multiplier,
    tolerance
  };
}

/**
 * The color descriptors.
 * @constant
 */
export default {
  black: descriptor("black", "#000000", 0, 0, null),
  brown: descriptor("brown", "#964B00", 1, 1, "±1%"),
  red: descriptor("red", "#FF0000", 2, 2, "±2%"),
  orange: descriptor("orange", "#FFA500", 3, 3, null),
  yellow: descriptor("yellow", "#FFFF00", 4, 4, "(±5%)"),
  green: descriptor("green", "#9ACD32", 5, 5, "±0.5%"),
  blue: descriptor("blue", "#6495ED", 6, 6, "±0.25%"),
  violet: descriptor("violet", "#9400D3", 7, 7, "±0.1%"),
  gray: descriptor("gray", "#A0A0A0", 8, 8, "±0.05%"),
  white: descriptor("white", "#FFFFFF", 9, 9, null),
  gold: descriptor("gold", "#CFB53B", null, -1, "±5%"),
  silver: descriptor("silver", "#C0C0C0", null, -2, "±10%"),
  pink: descriptor("pink", "#FF69B4", null, -3, null),
  transparent: descriptor("transparent", "#E9DDAF", null, null, "±20%")
};
