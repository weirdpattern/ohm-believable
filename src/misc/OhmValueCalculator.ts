import Resistance from "./Resistance";
import { descriptors, validateColor } from "./descriptors";
import {
  isMaybe,
  Maybe,
  BandColors,
  MultiplierColors,
  ToleranceColors
} from "./interfaces";

/**
 * Performs Ohm calculations based on the color of the bands.
 */
export interface OhmValueCalculator {
  /**
   * Calculates the Ohm value of a resistor based on the band colors.
   * @param {BandColors} aColor the color of the first band.
   * @param {BandColors} bColor the color of the second band.
   * @param {MultiplierColors} multiplierColor the color of the multiplier band.
   * @param {ToleranceColors} toleranceColor the color of the tolerance band.
   * @returns {Resistance} the resistance information.
   */
  CalculateOhmValue(
    aColor: BandColors,
    bColor: BandColors,
    multiplierColor: MultiplierColors,
    toleranceColor: ToleranceColors
  ): Resistance;
}

/**
 * Base implementation of the OhmValueCalculator interface.
 */
export class OhmValueCalculatorService implements OhmValueCalculator {
  /**
   * Calculates the Ohm value of a resistor based on the band colors.
   * @param {BandColors} aColor the color of the first band.
   * @param {BandColors} bColor the color of the second band.
   * @param {MultiplierColors} multiplierColor the color of the multiplier band.
   * @param {ToleranceColors} toleranceColor the color of the tolerance band.
   * @returns {Resistance} the resistance information.
   */
  public CalculateOhmValue(
    aColor: BandColors,
    bColor: BandColors,
    multiplierColor: MultiplierColors,
    toleranceColor: ToleranceColors
  ): Resistance {
    validateColor(aColor, "aColor");
    validateColor(bColor, "bColor");
    validateColor(multiplierColor, "multiplierColor");
    validateColor(toleranceColor, "toleranceColor");

    const [a, b, multiplier] = [
      descriptors[aColor].value,
      descriptors[bColor].value,
      descriptors[multiplierColor].multiplier
    ].map((value: Maybe<number>, index: number) => {
      const resolved: number = isMaybe(value) ? 0 : value;
      return index === 0 ? resolved * 10 : resolved;
    });

    return new Resistance(
      (a + b) * 10 ** multiplier,
      descriptors[toleranceColor].tolerance
    );
  }
}
