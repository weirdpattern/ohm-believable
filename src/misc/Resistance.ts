import { Maybe, isMaybe } from "../misc/interfaces";

/**
 * Resistance information.
 * @class
 */
export default class Resistance {
  /**
   * Class constructor.
   * @param {number} value the value of the resistance in ohms.
   * @param {string} [tolerance] the tolerance of the resistor.
   */
  public constructor(public value: number, public tolerance: Maybe<string>) {}

  /**
   * Gets a string representation of the resistance information.
   * @param {boolean} compact a flag indicating we want compact results.
   * @returns {string} the resistance information.
   */
  public toResistance(compact: boolean): string {
    let formattedValue: string = `${this.value}Ω`;

    if (compact) {
      if (this.value / 1000000 > 1) {
        formattedValue = `${(this.value / 1000000).toString()}MΩ`;
      } else if (this.value / 1000 > 1) {
        formattedValue = `${(this.value / 1000).toString()}KΩ`;
      }
    }

    let formattedTolerance = this.tolerance;
    if (isMaybe(this.tolerance)) {
      formattedTolerance = "";
    }

    return `${formattedValue} ${formattedTolerance}`;
  }
}
