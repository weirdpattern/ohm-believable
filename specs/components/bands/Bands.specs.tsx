import * as React from "react";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";

import Bands from "../../../src/components/bands";
import { ResistorBands } from "../../../src/misc/interfaces";

describe("Bands components", (): void => {
  it("renders correctly", (): void => {
    const bands: ResistorBands = {
      a: "yellow",
      b: "brown",
      multiplier: "violet",
      tolerance: "gold"
    };

    const tree: ReactTestRendererJSON = renderer
      .create(
        <Bands
          bands={bands}
          editingBand={null}
          onEditBand={null}
          onColorChange={null}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
