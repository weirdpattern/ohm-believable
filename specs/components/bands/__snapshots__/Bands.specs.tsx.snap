// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bands components renders correctly 1`] = `
<div
  className="resistor-svg"
>
  <svg
    height="60"
    version="1.1"
    viewBox="0 0 91.281248 15.875"
    width="345"
  >
    <g
      transform="translate(-43.65625,-115.19009)"
    >
      <g
        transform="translate(0.661135,-0.570325)"
      >
        <g>
          <path
            className="body"
            d="m 74.083332,117.08333 c -2.931904,1.0615 -6.614584,3.49644 -6.614583,6.61459 10e-7,3.11814 3.682681,5.55309 6.614583,6.61458 4.579779,1.65811 9.681387,-1.32292 14.552085,-1.32292 4.870697,0 9.972305,2.98103 14.552083,1.32292 2.9319,-1.06149 6.61458,-3.49644 6.61458,-6.61458 0,-3.11815 -3.68268,-5.55309 -6.61458,-6.61459 -4.579778,-1.65811 -9.681386,1.32292 -14.552083,1.32292 -4.870698,0 -9.972306,-2.98103 -14.552085,-1.32292 z"
          />
          <g>
            <path
              className="a clickable yellow"
              d="m 291.62305,441.57227 c -0.54239,-10e-4 -1.08405,0.009 -1.62305,0.0293 v 51.83985 c 3.27353,0.12476 6.61177,-0.16113 10,-0.68164 V 442.2832 c -2.263,-0.34701 -4.50307,-0.58983 -6.7168,-0.67578 -0.55486,-0.0216 -1.10863,-0.0341 -1.66015,-0.0351 z"
              onClick={[Function]}
              transform="scale(0.26458333)"
            />
            <path
              className="b clickable brown"
              d="m 310,444.3125 v 46.41602 c 3.3151,-0.77165 6.64765,-1.5728 10,-2.28321 V 446.5957 c -3.35231,-0.7104 -6.68493,-1.51157 -10,-2.2832 z"
              onClick={[Function]}
              transform="scale(0.26458333)"
            />
            <path
              className="multiplier clickable violet"
              d="m 330,448.21289 v 38.61523 c 1.66702,-0.15823 3.33418,-0.25 4.99805,-0.25 1.66516,0 3.33326,0.0915 5.00195,0.25 v -38.61523 c -1.66869,0.15847 -3.33679,0.25195 -5.00195,0.25195 -1.66387,0 -3.33103,-0.0937 -4.99805,-0.25195 z"
              onClick={[Function]}
              transform="scale(0.26458333)"
            />
            <path
              className="tolerance clickable gold"
              d="m 378.37695,441.57031 c -2.75154,0.005 -5.54678,0.27595 -8.37695,0.71094 v 50.47852 c 3.38852,0.52078 6.72652,0.8066 10,0.68164 v -51.8418 c -0.53939,-0.0206 -1.08021,-0.0303 -1.62305,-0.0293 z"
              onClick={[Function]}
              transform="scale(0.26458333)"
            />
          </g>
        </g>
        <path
          className="line"
          d="m 109.80273,123.44727 v 0.5 h 23.8125 v -0.5 z"
        />
        <path
          className="line"
          d="m 43.65625,123.44727 v 0.5 h 23.8125 v -0.5 z"
        />
      </g>
    </g>
  </svg>
</div>
`;
